File: src/Exec.php
<?php

namespace xltxlm\shell;

use xltxlm\logger\thelostlog_exec\thelostlog_exec;
use xltxlm\shell\Exception\Exception_Exec;
use xltxlm\shell\Exec\Exec_implements;


/**
 * 执行系统命令,错误会抛出异常;
 */
class Exec extends Exec\Exec_implements
{

    /**
     * Exec constructor.
     */
    public function __construct(string $cmd = '')
    {
        if ($cmd) {
            $this->setCmd($cmd);
        }
    }

    /**
     * @param \Redis $RedisObject ;
     * @return $this
     */
    public static function setRedisObject(\Redis $RedisObject)
    {
        Exec::$RedisObject = $RedisObject;
        if (is_object(Exec::$RedisObject) && $_SERVER['logid']) {
            Exec::$RedisObject->zAdd($_SERVER['logid'], Exec::getcounti(),'Begin');
            //保存10分钟
            Exec::$RedisObject->expire($_SERVER['logid'], 60 * 10);
        }
    }

    /**
     * @return int;
     */
    public static function getcounti(): int
    {
        Exec::$counti++;
        return Exec::$counti;
    }


    public function __invoke(): string
    {
        //如果声明了Redis存储,那么每一个命令都会给存储,保存10分钟
        if (is_object(Exec::$RedisObject) && $_SERVER['logid']) {
            Exec::$RedisObject->zAdd($_SERVER['logid'], Exec::getcounti(),$this->getCmd());
        }
        $thelostlog_exec = (new thelostlog_exec())
            ->setcommand($this->getCmd());
        $start = microtime(true);
        exec($this->getCmd(), $out, $return);
        $time = sprintf('%.4f', microtime(true) - $start);
        $ressult = join("\n", $out);
        if ($this->isDebug()) {
            p(["cmd" => $this->getCmd(), "time" => $time, 'ressult' => $ressult, 'return' => $return]);
        }
        if (is_object(Exec::$RedisObject) && $_SERVER['logid']) {
            Exec::$RedisObject->zAdd($_SERVER['logid'], Exec::getcounti(),"return code:$return");
        }
        $thelostlog_exec->setresult($ressult);
        if ($this->isException_on_Fail() && $return) {
            $thelostlog_exec->seterror($return);
            throw new Exception_Exec("执行命令[{$this->getCmd()}]错误:" . join($out) . ",错误值:$return");
        }
        return $ressult;
    }


}

File: go/Exec.go
package 

type Exec struct {
    /* 要执行的命令 */
    Cmd string

    /* 是否在php错误文件输出执行的命令 */
    Debug bool

    /* 错误的时候抛出异常信息 */
    Exception_on_Fail bool

    /* 存储全部执行过程命令的Redis服务器 */
    RedisObject 

    /* 记录的顺序 */
    counti int

}

func NewExec(Cmd string,Debug bool,Exception_on_Fail bool,RedisObject ) *Exec{
    var this = new(Exec)
    this.SetCmd(Cmd);
    this.SetDebug(Debug);
    this.SetException_on_Fail(Exception_on_Fail);
    this.SetRedisObject(RedisObject);
    return this
}

func (this *Exec)GetCmd() string{
    return this.Cmd;
}

func (this *Exec)SetCmd(Cmd string) *Exec{
    this.Cmd = Cmd;
    return this
}
func (this *Exec)GetDebug() bool{
    return this.Debug;
}

func (this *Exec)SetDebug(Debug bool) *Exec{
    this.Debug = Debug;
    return this
}
func (this *Exec)GetException_on_Fail() bool{
    return this.Exception_on_Fail;
}

func (this *Exec)SetException_on_Fail(Exception_on_Fail bool) *Exec{
    this.Exception_on_Fail = Exception_on_Fail;
    return this
}
func (this *Exec)GetRedisObject() {
    return this.RedisObject;
}

func (this *Exec)SetRedisObject(RedisObject ) *Exec{
    this.RedisObject = RedisObject;
    return this
}

/**
    返回命令的内容*/
func (this *Exec)__invoke()string{

}

